src/algebra/order/pointwise.lean

```diff
@@ -1,20 +1,276 @@
+
+lemma Inf_image2_eq_Inf_Sup (h₁ : ∀ b, galois_connection (l₁ b) (swap u b))
+  (h₂ : ∀ a, galois_connection (to_dual ∘ l₂ a) (u a ∘ of_dual)) :
+  Inf (image2 u s t) = u (Inf s) (Sup t) :=
+@Inf_image2_eq_Inf_Inf _ (order_dual β) _ _ _ _ _ _ _ _ _ h₁ h₂
+
+lemma Inf_image2_eq_Sup_Inf (h₁ : ∀ b, galois_connection (to_dual ∘ l₁ b) (swap u b ∘ of_dual))
+  (h₂ : ∀ a, galois_connection (l₂ a) (u a)) :
+  Inf (image2 u s t) = u (Sup s) (Inf t) :=
+@Inf_image2_eq_Inf_Inf (order_dual α) _ _ _ _ _ _ _ _ _ _ h₁ h₂
+
+lemma Inf_image2_eq_Sup_Sup (h₁ : ∀ b, galois_connection (to_dual ∘ l₁ b) (swap u b ∘ of_dual))
+  (h₂ : ∀ a, galois_connection (to_dual ∘ l₂ a) (u a ∘ of_dual)) :
+  Inf (image2 u s t) = u (Sup s) (Sup t) :=
+@Inf_image2_eq_Inf_Inf (order_dual α) (order_dual β) _ _ _ _ _ _ _ _ _ h₁ h₂
+
+end
+
+namespace set
+variables {α β γ : Type*} {f : α × β → γ} {s : set α} {t : set β}
+
+lemma image_prod' : f '' s ×ˢ t = image2 (curry f) s t :=
+by { rw [←image_prod (curry f), ←uncurry_curry f], refl }
+
+end set
+
+open set
+
+section
+variables {α β γ : Type*} [conditionally_complete_lattice α] [conditionally_complete_lattice β]
+  [conditionally_complete_lattice γ] {f : α → β → γ} {s : set α} {t : set β}
+
+open order_dual
+
+lemma csupr_le_iff {ι : Sort*} [nonempty ι] {f : ι → α} {a : α} (hf : bdd_above (range f)) :
+  supr f ≤ a ↔ ∀ i, f i ≤ a :=
+(is_lub_le_iff $ is_lub_csupr hf).trans forall_range_iff
+
+lemma le_cinfi_iff {ι : Sort*} [nonempty ι] {f : ι → α} {a : α} (hf : bdd_below (range f)) :
+  a ≤ infi f ↔ ∀ i, a ≤ f i :=
+(le_is_glb_iff $ is_glb_cinfi hf).trans forall_range_iff
+
+lemma bcsupr_le_iff {ι : Type*} {s : set ι} {f : ι → α} {a : α} (hs : s.nonempty)
+  (hf : bdd_above (f '' s)) :
+  (⨆ i : s, f i) ≤ a ↔ ∀ i ∈ s, f i ≤ a :=
+(is_lub_le_iff $ is_lub_csupr_set hf hs).trans ball_image_iff
+
+lemma le_bcinfi_iff {ι : Type*} {s : set ι} {f : ι → α} {a : α} (hs : s.nonempty)
+  (hf : bdd_below (f '' s)) :
+  a ≤ (⨅ i : s, f i) ↔ ∀ i ∈ s, a ≤ f i :=
+(le_is_glb_iff $ is_glb_cinfi_set hf hs).trans ball_image_iff
+
+lemma bcsupr_prod (f : α × β → γ) :
+  (⨆ x : (s ×ˢ t : set (α × β)), f x) = ⨆ (a : s) (b : t), f (a, b) := sorry
+
+lemma bcinfi_prod (f : α × β → γ) :
+  (⨅ x : (s ×ˢ t : set (α × β)), f x) = ⨅ (a : s) (b : t), f (a, b) :=
+eq_of_forall_le_iff $ λ c, begin
+  have hs : s.nonempty := sorry,
+  have ht : t.nonempty := sorry,
+  have hf : bdd_below (f '' s ×ˢ t) := sorry,
+  rw [le_bcinfi_iff (hs.prod ht) hf, prod.forall],
+  refine (forall_congr $ λ a, _).trans
+    (@le_bcinfi_iff _ _ _ _ (λ a, ⨅ (b : ↥t), f (a, b)) c hs _).symm,
+  simp_rw [mem_prod, and_imp],
+  rw forall_swap,
+  refine forall_congr (λ ha, _),
+  rw [@le_bcinfi_iff _ _ _ _ (λ b, f (a, b)) _ ht],
+  rw image_prod' at hf,
+  exact hf.mono (image_subset_image2_right ha),
+  refine hf.mono _,
+  sorry
+end
+
+lemma cSup_image2 : Sup (image2 f s t) = ⨆ (a : s) (b : t), f a b :=
+by { rw [←image_prod, Sup_image'], exact bcsupr_prod (uncurry f) }
```

Comment:
How would you prove this, @eric-wieser ? Is this a reasonable lemma to have (with maybe some boundedness and nonemptiness assumptions)?

Reply:
I would be tempted to prove this by rewriting by `set.image2_eq_Union` from my map PE

Reply:
And how do you overcome the fact that you need to prove something about the boundedness of `range (λ a, ⨆ (b : t), f a b)`? This is what I'm hitting in the proof of `bcinfi_prod` just above.

Reply:
I can't really make sense of it, but using `cSup_image2` was a wrong turn. I understand that it's false, but I don't see how the replacement I found is fundamentally different from it.